icaruscode/PMT/OpReco/BarycenterMatchProducer_module.cc

```diff
@@ -0,0 +1,548 @@
+
+//ROOT includes
+#include "TTree.h"
+#include "TVector3.h"
+
+#include <memory>
+
+using microseconds = util::quantities::intervals::microseconds;
+using electronics_time = detinfo::timescales::electronics_time;
+
+class BarycenterMatchProducer;
+
+
+class BarycenterMatchProducer : public art::EDProducer {
+public:
+  explicit BarycenterMatchProducer(fhicl::ParameterSet const& p);
+  // The compiler-generated destructor is fine for non-base
+  // classes without bare pointers or other resource use.
+
+  // Plugins should not be copied or assigned.
+  BarycenterMatchProducer(BarycenterMatchProducer const&) = delete;
+  BarycenterMatchProducer(BarycenterMatchProducer&&) = delete;
+  BarycenterMatchProducer& operator=(BarycenterMatchProducer const&) = delete;
+  BarycenterMatchProducer& operator=(BarycenterMatchProducer&&) = delete;
+
+  // Required functions.
+  void produce(art::Event& e) override;
+
+private:
+
+  // Declare member data here.
+  void InitializeSlice();                                                                                  ///< Re-initialize all slice-level data members
+  double CentroidOverlap(double center1, double center2, double width1, double width2);                    ///< Return overlap between charge and light centroids OR distance apart if no overlap
+  double CalculateAsymmetry(art::Ptr<recob::OpFlash> flash, int cryo);                                     ///< Return the east-west asymmetry of PEs in a given OpFlash
+  void updateChargeVars(double sumCharge, TVector3 sumPos, TVector3 sumPosSqr, double triggerFlashCenter); ///< Update slice-level data members with charge and trigger match info
+  void updateFlashVars(art::Ptr<recob::OpFlash> flash, double firstHit);                     ///< Update slice-level data members with best match info
+  void updateMatchInfo(sbn::BarycenterMatch& matchInfo);                                                   ///< Update match product with slice-level data members
+ 
+  // Input parameters
+  std::vector<std::string>  fInputTags;            ///< Suffix added onto fOpFlashLabel and fPandoraLabel, used by ICARUS for separate cryostat labels but could be empty
+  std::string               fOpFlashLabel;         ///< Label for PMT reconstruction products
+  std::string               fPandoraLabel;         ///< Label for Pandora output products
+  std::string               fTriggerLabel;         ///< Label for trigger product
+  bool                      fCollectionOnly;       ///< Only use TPC spacepoints from the collection plane
+  bool                      fUseTimeRange;         ///< Reject impossible matches based on allowed time range of TPC hits relative to trigger 
+  bool                      fVerbose;              ///< Print extra info
+  bool                      fFillMatchTree;        ///< Fill an output TTree in the supplemental file
+  double                    fNominalTrigTime;      ///< Typical time of triggering flash, EYEBALLED (us)
```

Comment:
Can we flip the sign of this and call it `fTriggerDelay`? at least in data, the main reason of the fact that the trigger is after the flash is that it takes a while to the trigger hardware to spit out a response.
Ok, it's not just that... but overall it may be interpreted as a trigger delay rather than a "nominal" trigger time (in fact, `1500.0` is more nominal than `1499.78`).